src/main/java/gregtech/api/unification/material/Material.java

```diff
@@ -0,0 +1,848 @@
+        this.flags = flags;
+
+        this.properties.setMaterial(this);
+        registerMaterial(this);
+    }
+
+    // thou shall not call
+    protected Material(String name) {
+        materialInfo = new MaterialInfo(0, name);
+        materialInfo.iconSet = MaterialIconSet.NONE;
+        properties = new MaterialProperties();
+        flags = new MaterialFlags();
+    }
+
+    protected void registerMaterial(Material material) {
+        MaterialRegistry.register(this);
+    }
+
+    public void addFlag(MaterialFlag... flags) {
+        if (MaterialRegistry.isFrozen())
+            throw new IllegalStateException("Cannot add flag to material when registry is frozen!");
+        this.flags.addFlags(flags).verify(this);
+    }
+
+    //@ZenMethod("hasFlagRaw")
+    public boolean hasFlag(MaterialFlag flag) {
+        return flags.hasFlag(flag);
+    }
+
+    public boolean hasFlags(MaterialFlag... flags) {
+        return Arrays.stream(flags).allMatch(this::hasFlag);
+    }
+
+    //@ZenMethod
+    //public boolean hasFlag(String flagName) {
+    //    long materialFlagId = MaterialFlags.resolveFlag(flagName, getClass());
+    //    return hasFlag(materialFlagId);
+    //}
+
+    protected void calculateDecompositionType() {
+        if (!materialInfo.componentList.isEmpty() &&
+            !hasFlag(MaterialFlags.DECOMPOSITION_BY_CENTRIFUGING) &&
+            !hasFlag(MaterialFlags.DECOMPOSITION_BY_ELECTROLYZING) &&
+            !hasFlag(MaterialFlags.DISABLE_DECOMPOSITION)) {
+            boolean onlyMetalMaterials = true;
+            for (MaterialStack materialStack : materialInfo.componentList) {
+                Material material = materialStack.material;
+                onlyMetalMaterials &= material.hasProperty(PropertyKey.INGOT);
+            }
+            //allow centrifuging of alloy materials only
+            if (onlyMetalMaterials) {
+                flags.addFlags(MaterialFlags.DECOMPOSITION_BY_CENTRIFUGING);
+            } else {
+                flags.addFlags(MaterialFlags.DECOMPOSITION_BY_ELECTROLYZING);
+            }
+        }
+    }
+
+    public Fluid getFluid() {
+        FluidProperty prop = getProperty(PropertyKey.FLUID);
+        if (prop == null)
+            throw new IllegalArgumentException("Material " + materialInfo.name + " does not have a Fluid!");
+        return prop.getFluid();
+    }
+
+    public FluidStack getFluid(int amount) {
+        return new FluidStack(getFluid(), amount);
+    }
+
+    public int getHarvestLevel() {
+        if (!hasProperty(PropertyKey.DUST))
+            throw new IllegalArgumentException("Material " + materialInfo.name + " does not have a harvest level! Is probably a Fluid");
+        return getProperty(PropertyKey.DUST).getHarvestLevel();
+    }
+
+    //@ZenMethod
+    public void setMaterialRGB(int materialRGB) {
+        materialInfo.color = materialRGB;
+    }
+
+    //@ZenGetter
+    public int getMaterialRGB() {
+        return materialInfo.color;
```

Comment:
Could we do something similar to chemical formula here? I.e. we take average rgb of all material components. This is something we do manually a lot with SimpleDustMaterials